### PR TITLE
[FLINK-15639][k8s] Support to set tolerations for jobmanager and taskmanger pod

### PR DESCRIPTION
**Pull request by wangyang0918**
## What is the purpose of the change

The [toleration](https://kubernetes.io/docs/concepts/configuration/taint-and-toleration/) is used to separate the K8s cluster into several individual partitions. Usually it it related with business group. So i treat is as a first class feature. Since at least in our production environment, every Flink job need to specify the toleration so that it could be scheduled to the corresponding resource pool.

Moreover, it is sth very like YARN partition.

## Brief change log

* Introduce the config option and set toleration to pod

## Verifying this change

* Covered by unit test
* Manually test in a real k8s cluster

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: (yes)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (yes)
  - If yes, how is the feature documented? (docs)

### CODE REVIEW

**Comment by wangyang0918** [comment]
@TisonKun @zhengcanbin Do you mind to take a look at your convenience?

File: flink-kubernetes/src/main/java/org/apache/flink/kubernetes/kubeclient/resources/KubernetesToleration.java [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+
+	private static final Logger LOG = LoggerFactory.getLogger(KubernetesToleration.class);
+
+	private KubernetesToleration(Toleration toleration) {
+		super(toleration);
+	}
+
+	public static KubernetesToleration fromMap(Map<String, String> stringMap) {
+		final TolerationBuilder tolerationBuilder = new TolerationBuilder();
+		stringMap.forEach((k, v) -> {
```
Comment (zhengcanbin) [inline flink-kubernetes/src/main/java/org/apache/flink/kubernetes/kubeclient/resources/KubernetesToleration.java:41]: Could the keys be case insensitive?

File: flink-kubernetes/src/main/java/org/apache/flink/kubernetes/kubeclient/parameters/KubernetesParameters.java [hunk 2]
```diff
@@ -62,6 +63,12 @@
 	 */
 	Map<String, String> getAnnotations();
 
+	/**
+	 * A collection of tolerations that are set to the JobManager and TaskManager Pod(s). Kubernetes Taints and
```
Comment (zhengcanbin) [inline flink-kubernetes/src/main/java/org/apache/flink/kubernetes/kubeclient/parameters/KubernetesParameters.java:67]: nit: should be taints?

Comment (wangyang0918) [reply flink-kubernetes/src/main/java/org/apache/flink/kubernetes/kubeclient/parameters/KubernetesParameters.java:67]: I will fix it.

File: flink-kubernetes/src/main/java/org/apache/flink/kubernetes/kubeclient/resources/KubernetesToleration.java (hunk 1 above)
Comment (wangyang0918) [reply flink-kubernetes/src/main/java/org/apache/flink/kubernetes/kubeclient/resources/KubernetesToleration.java:41]: Thanks for your suggestion. I think it makes sense to ignore character case here.

**Comment by wangyang0918** [comment]
@zhengcanbin I have integrated your suggestions and rebased the latest master. Please take another look at your convenience.

**Comment by wangyang0918** [comment]
@TisonKun Do you mind to also take a look? I think the changes in this PR is quite straightforward.

File: flink-kubernetes/src/main/java/org/apache/flink/kubernetes/configuration/KubernetesConfigOptions.java [hunk 3]
```diff
@@ -186,6 +186,24 @@
 		.withDescription("The user-specified annotations that are set to the TaskManager pod. The value could be " +
 			"in the form of a1:v1,a2:v2");
 
+	public static final ConfigOption<List<Map<String, String>>> JOB_MANAGER_TOLERATIONS =
+		key("kubernetes.jobmanager.tolerations")
+			.mapType()
+			.asList()
+			.noDefaultValue()
+			.withDescription("The user-specified tolerations to be set to the JobManager pod. The value could be " +
```
Comment (TisonKun) [inline flink-kubernetes/src/main/java/org/apache/flink/kubernetes/configuration/KubernetesConfigOptions.java:194]:
```suggestion
			.withDescription("The user-specified tolerations to be set to the JobManager pod. The value should be " +
```

File: flink-kubernetes/src/main/java/org/apache/flink/kubernetes/configuration/KubernetesConfigOptions.java [hunk 4]
```diff
@@ -186,6 +186,24 @@
...
+				"in the form of key:key1,operator:Equal,value:value1,effect:NoSchedule;" +
+				"key:key2,operator:Exists,effect:NoExecute,tolerationSeconds:6000");
+
+	public static final ConfigOption<List<Map<String, String>>> TASK_MANAGER_TOLERATIONS =
+		key("kubernetes.taskmanager.tolerations")
+			.mapType()
+			.asList()
+			.noDefaultValue()
+			.withDescription("The user-specified tolerations to be set to the TaskManager pod. The value could be " +
```
Comment (TisonKun) [inline flink-kubernetes/src/main/java/org/apache/flink/kubernetes/configuration/KubernetesConfigOptions.java:203]:
```suggestion
			.withDescription("The user-specified tolerations to be set to the TaskManager pod. The value should be " +
```

File: flink-kubernetes/src/main/java/org/apache/flink/kubernetes/configuration/KubernetesConfigOptions.java (hunk 3 above)
Comment (wangyang0918) [reply flink-kubernetes/src/main/java/org/apache/flink/kubernetes/configuration/KubernetesConfigOptions.java:194]: Make sense. I will fix it.

**Comment by wangyang0918** [comment]
@TisonKun Thanks for your review. I have integrated your comments and force pushed the PR. Please have another look.

**Pull request closed without merging**